Refactoring message to sign for web3 authentication

silva-fj opened the pull request:
As topic, this PR updates and simplifies the web3 message to sign as well as the signature verification. A new RPC method has been added to get the message payload to sign.

The `verify_auth` function no longer requires the `NativeTaskWrapper`.

This is part of the omni-executor's auth rework.

Note: This PR only includes changes for web3 authentication, the `AddAccount` (to AccountStore) native call still uses the old format for the validation data

silva-fj committed:
feat(auth): implement Web3 authentication message codes

Add time-based authentication message code generator for Web3
authentication flows.

silva-fj committed:
refactor(auth): simplify Web3 authentication flow

Replace complex nonce-based authentication with a time-based message
code system for Web3 authentication.

silva-fj committed:
test(auth): add unit test for Web3 authentication

silva-fj committed:
feat(rpc): add omni_getMessageCode RPC method

Introduce the `omni_getMessageCode` RPC method to generate and return a
message code for web3 authentication.

silva-fj committed:
feat(web3-auth): use message code for task signing

Refactored integration tests and type creators to use a dynamic message
code for signing native tasks in Web3 authentication flows.

silva-fj committed:
refactor(rpc): remove mrenclave from server context

silva-fj committed:
fixing typo

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Update the `verify_auth` function to accept explicit auth and
sender parameters, removing the need to pass the entire wrapper.

kziemianek commented on tee-worker/omni-executor/heima/authentication/src/web3.rs:
```diff
@@ -0,0 +1,52 @@
+/// # Returns
+///
+/// A string containing a 6-digit message code that changes every `period` seconds
+pub fn generate_message_code(period_in_seconds: u64) -> String {
```
This function could take `now` as argument so there would be no need to wait in tests.

silva-fj replied:
I've adjusted the code 👍🏼

kziemianek commented on tee-worker/omni-executor/heima/authentication/src/web3.rs:
```diff
@@ -0,0 +1,52 @@
+		assert_eq!(code.len(), 6);
+
+		// Generate the code within the same period
+		sleep(Duration::new(2, 0));
```
This fails ocassionaly. If you change sleep time to 4 it will fail more often.

silva-fj committed:
refactor: improve web3 msg code testability

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

kziemianek commented on tee-worker/omni-executor/heima/authentication/src/web3.rs:
```diff
@@ -0,0 +1,52 @@
+	fn test_message_code_at() {
+		let period = 5;
+
+		let base_ts = 1_000;
```
This test fail if you use current ts `let base_ts = 1745322204;`

silva-fj replied:
Yeah that's because when dividing by that period it gets at the edge of the next window, so just adding 2 secs will make it jump to the next. The function generates the code for the time window using the period regardless of when the first code was requested (just like google authenticator for example). I guess we could return from the RPC method the code + time left for the next one to simulate authenticator apps. Otherwise well have to persist some sort of records to keep track of each code

silva-fj replied:
I've updated the RPC method to return the time left until the next code

kziemianek replied:
Ok I see so in some cases this method will fail (for example if it takes too long for request to reach to server).

silva-fj committed:
feat(auth): return seconds left with message code

Update the message code generation to return both the code and the number
of seconds left in the current period. Adjust the RPC response and
verification logic to accommodate this change. Add tests for period
boundary behavior and seconds left calculation.

silva-fj committed:
updating integration tests

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj commented on tee-worker/omni-executor/heima/authentication/src/web3.rs:
```diff
@@ -0,0 +1,91 @@
+	pub message_code: String,
+}
+
+pub const MESSAGE_CODE_PERIOD: u64 = 60; // 1 minute
```
What would be a sensible value here?

silva-fj replied:
Not relevant anymore

silva-fj committed:
feat(core): add Hashable trait and impl for AccountId

silva-fj committed:
feat(helpers): make generate_otp function public

silva-fj committed:
feat(rpc): redesign web3 message code generation and verify

Refactor the omni_getMessageCode RPC to generate and store a unique OTP
message code per omni account, replacing the previous static code approach.
Update web3 authentication verification to fetch and validate the stored code
from persistent storage. Unify verification error handling and update tests
to reflect the new logic.

silva-fj committed:
test(omni-account): update getMessageCode usage in tests

Refactor integration tests to pass the omniAccount parameter to
getMessageCode, aligning with the updated function signature. This ensures
tests correctly request message codes for specific omni accounts and match
the new RPC API requirements.

silva-fj committed:
refactor(auth): remove chrono dependency and time-based code

Eliminate the unused chrono dependency and related time-based message code
generation logic from the authentication module.

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
fixing dev-dependencies

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
fixing merging issues

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj commented:
I'll include these changes in another PR
